### PR TITLE
[Doc] Revise "Why a new programming language"

### PR DESCRIPTION
**Pull request by k-ye**
Related issue = #

Preview: https://deploy-preview-4306--docsite-preview.netlify.app/lang/articles/basic/overview

### CODE REVIEW

File: docs/lang/articles/basic/overview.md [hunk 1]
```diff
@@ -2,22 +2,48 @@
...
+
+The concept of decoupling is further extended to the type system. With GPU memory capacity and bandwidth becoming the major bottlenecks nowadays, it is vital to be able to pack more data per memory unit. Since 2021, Taichi has introduced customizable quantized types, allowing for the definition of fixed point or floating point numbers with arbitrary bits (still needs to be under 64). This has allowed an MPM simulation of over 400 million particles on a single GPU device. Learn more details in [the QuanTaichi paper](https://yuanming.taichi.graphics/publication/2021-quantaichi/quantaichi.pdf).
+
+Taichi is intuitive. If you know Python, you know Taichi. If you write Taichi, you awaken your GPU (or CPU as a fallback). Ever since its debut, this simple idea has gained so much popularity, that many were attracted to contribute new backends, including Vulkan, OpenGL and DirectX (working in progress). Without our strong and dedicated community, Taichi would never have been where it is now.
+
+Going forward, we see many new opportunities lying ahead this path, and would like to share some of our vision with you.
+
+**Academia**
```
Comment (k-ye) [inline docs/lang/articles/basic/overview.md:28]: Additional input regarding this part :-)? @ltt1598

File: docs/lang/articles/basic/overview.md [hunk 2]
```diff
@@ -2,22 +2,48 @@
...
+
+Taichi will keep embracing the academia. Towards this goal, we have implemented or planned for features that are particular useful for researchers, small matrix and sparse matrix solvers, to name a few. Taichi has also supported automatic differentiation, a key feature for differential simulation and robotics training.
```
Comment (turbo0628) [inline docs/lang/articles/basic/overview.md:30]: differential -> differentiable

File: docs/lang/articles/basic/overview.md [hunk 3]
```diff
@@ -2,22 +2,48 @@
...
+![](https://github.com/taichi-dev/taichi_assets/blob/master/static/imgs/unity_fluid.gif?raw=true)
+
+**General-purpose computing**
+
+While originally designed for physics simulation, Taichi has found its application in many other areas that can be boosted by GPU  general-purpose computing.
+
+* [taichimd](https://github.com/victoriacity/taichimd): Interactive, GPU-accelerated Molecular (& Macroscopic) Dynamics using the Taichi programming language
+* [TaichiSLAM](https://github.com/xuhao1/TaichiSLAM): a 3D Dense mapping backend library of SLAM based Taichi-Lang, designed for the aerial swarm.
+* [Stannum](https://github.com/ifsheldon/stannum): Fusing Taichi into PyTorch.
+
```
Comment (turbo0628) [inline docs/lang/articles/basic/overview.md:46]: Maybe we can also mention the fluid solver here? https://github.com/hietwll/LBM_Taichi

File: docs/lang/articles/basic/overview.md (hunk 2 above)
Comment (k-ye) [reply docs/lang/articles/basic/overview.md:30]:
```suggestion
Taichi will keep embracing the academia. Towards this goal, we have implemented or planned for features that are particular useful for researchers, small matrix and sparse matrix solvers, to name a few. Taichi has also supported automatic differentiation, a key feature for differentiable simulation and robotics training.
```

File: docs/lang/articles/basic/overview.md (hunk 3 above)
Comment (k-ye) [reply docs/lang/articles/basic/overview.md:46]: I was hoping to use some cases that are not simulation LOL

**Comment by turbo0628** [comment]
I truly enjoy reading this article, LGTM!

File: docs/lang/articles/basic/overview.md (hunk 1 above)
Comment (ltt1598) [reply docs/lang/articles/basic/overview.md:28]: I think we can echo the high productivity using Taichi at first: 90% of the research code will be trashed due to the nature of research where assumptions keep being broken and ideas keep being iterated. Swiftly coding without thinking too much about performance may lead to incorrect conclusions, while pre-matured code optimization can be a waste of time and often produces a tangled mess. The high performance and productivity are, therefore, extremely helpful for research projects.

The key features we have (or plan to have) for high-performance computing research projects include small-scale linear algebra (inside kernels), large-scale sparse systems, and efficient neighbor accessing for both structured and unstructured data.

We also provide an automatic differentiation module via source code transformation (at IR level), making Taichi a sweet differentiable simulation tool for machine learning projects.

**Comment by ltt1598** [comment]
I like the new "why new programming language" page a lot :)

**Comment by Vissidarte-Herman** [comment]
Never in my life have I read something its equal. Inspiring, enlightening, and a breath of fresh air!

File: docs/lang/articles/basic/overview.md (hunk 1 above)
Comment (k-ye) [reply docs/lang/articles/basic/overview.md:28]: Truly appreciated!

**Comment by k-ye** [comment]
Three warm compliments without a single LGTM, I'm heartbroken...

**Comment by ltt1598** [comment]
> 

LGTM now :)

**Review by strongoier — APPROVED** [review]
A green LGTM!

File: docs/lang/articles/basic/overview.md [hunk 4]
```diff
@@ -2,22 +2,48 @@
...
-- Productivity
-- Performance
-- Portability
-- Spatially sparse computation
-- Differentiable programming
-- Metaprogramming
+If all these sound too familiar to you, congratulations, you are probably looking at the right solution.
 
-## Design decisions
+Originally born from the MIT CSAIL lab, Taichi was designed to facilitate computer graphics researchers' day-to-day life, by helping them quickly implement visual computing and physics simulation algorithms that are executable on GPU. The path Taichi took was an innovative one: Taichi is embedded in Python, and utilizes the modern just-in-time (JIT) frameworks (e.g., LLVM, SPIR-V) to offload the Python source code to native GPU or CPU instructions, offering the performance at both development time and runtime.
```
[the remark this answers is not in the document]
Comment (k-ye) [reply docs/lang/articles/basic/overview.md:11]: Thanks, I tweak some, but prefer to keep "The path Taichi took aas ..."

File: docs/lang/articles/basic/overview.md [hunk 5]
```diff
@@ -2,22 +2,48 @@
...
+4. JIT compilation provides additional optimization opportunities.
+
+That said, Taichi is much more than just a Python JIT transpiler. One of the initial design goals is to *decouple the computation from the data structures*. The mechanism Taichi provides is a set of generic data containers, called *SNode* (/ˈsnoʊd/). SNodes can be used to compose hierarchical, dense or sparse, multi-dimensional fields conveniently. Switching between array-of-structures and structure-of-arrays layouts is usually a matter of ≤10 lines of code. This has truly enabled many use cases in numerical simulation. If you are interested to learn them, please check out [Fields (advanced)](https://docs.taichi.graphics/lang/articles/advanced/layout), [Sparse spatial data structures](https://docs.taichi.graphics/lang/articles/advanced/sparse), or [the original Taichi paper](https://yuanming.taichi.graphics/publication/2019-taichi/taichi-lang.pdf).
```
[the remark this answers is not in the document]
Comment (k-ye) [reply docs/lang/articles/basic/overview.md:20]: > Be interested in doing

Good point, but then I actually searched for that when writing this line...

https://jakubmarian.com/interested-in-doing-vs-interested-to-do-in-english/

File: docs/lang/articles/basic/overview.md [hunk 6]
```diff
@@ -2,22 +2,48 @@
...
 
-- Decouple computation from data structures
-- Domain-specific compiler optimizations
-- Megakernels
-- Two-scale automatic differentiation
-- Embedding in Python
+To be fair, a domain-specific language (DSL) with a Python frontend is not something new. In the past few years, frameworks like Halide, PyTorch and TVM have grown ripe and become the de facto standards in areas like image processing and deep learning (DL). What differentiates Taichi the most from these frameworks is its imperative programming paradigm. Being a DSL, Taichi isn't so specialized in a particular computing pattern, thus provides better flexibility. While one may argue that flexibility usually comes at a cost of not being fully optimized, we find this is often not the case for a few reasons.
```
[the remark this answers is not in the document]
Comment (k-ye) [reply docs/lang/articles/basic/overview.md:13]: Sweet, I believe it should be "we often find this not to be the case"?

File: docs/lang/articles/basic/overview.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (Vissidarte-Herman) [reply docs/lang/articles/basic/overview.md:20]: You raised a good point! Go with it!

File: docs/lang/articles/basic/overview.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (k-ye) [reply docs/lang/articles/basic/overview.md:13]: I see, sounds good!

**Comment by k-ye** [comment]
Lacking content: 

* difference between Taichi and GLSL
* btw Taichi and physics engine
* what scenarios should you use/not use Taichi

**Pull request merged**